Add Dockerfile

OddMagnet opened the pull request:
Saw someone on reddit requesting a docker image for this and I always wanted to try and create a Dockerfile for something, so I thought I'd try it for this project :D

Everything seems to run fine with it, thought I'm no expert on creating Dockerfiles

The next step would be to [build & publish](https://docs.docker.com/get-started/docker-concepts/building-images/build-tag-and-publish-an-image/), either manually or via github actions I guess, though I have not experience with that yet

Note: I wasn't able to get only `scanadf`, always got the error that the package couldn't be found, so I used `sane`, which includes `scanadf`

OddMagnet committed:
Add Dockerfile

nuvious commented:
Good comment, sane is the right package so I'll adjust the documentation accordingly or you're more than welcome to correct it in your PR. I can also setup the GitHub action because I've done that before and can publish it to my docker hub.

nuvious commented on Dockerfile:
```diff
@@ -0,0 +1,7 @@
+WORKDIR /
+COPY . .
+RUN pip3 install Flask
+RUN apt update && apt install -y sane ocrmypdf
```
Move the installation of the dependencies above the WORKDIR statement. When docker builds it doesn't repeat a layer build if nothing about that layer or the layers before it has changed. By copying the code before installing the dependencies, any change in code results in a re-build of the dependency installation layer which eats up time during development.

Minor thing but saves a lot of time in rapid-development in docker environments.

nuvious commented on Dockerfile:
```diff
@@ -0,0 +1,7 @@
+RUN pip3 install Flask
+RUN apt update && apt install -y sane ocrmypdf
+EXPOSE 5000
+CMD ["python3", "./app.py"]
```
This is a bit of an obscure issue, but the script fails on a policy violation with ImageMagick-6 when run in a docker container and was failing for me in testing:

```bash
+ convert image-0001 2024-12-23-05-05-35-test.pdf
convert-im6.q16: attempt to perform an operation not allowed by the security policy `PDF' @ error/constitute.c/IsCoderAuthorized/426.
```

Simple fix, we just need to comment out:

```xml
    <policy domain="coder" rights="none" pattern="PDF" />
```

in `/etc/ImageMagick-6/policy.xml`. An easier way to manage this is to get the full policy file, edit it and add it. Below is the file itself with that line commented out:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE policymap [
  <!ELEMENT policymap (policy)*>
  <!ATTLIST policymap xmlns CDATA #FIXED ''>
  <!ELEMENT policy EMPTY>
  <!ATTLIST policy xmlns CDATA #FIXED '' domain NMTOKEN #REQUIRED
    name NMTOKEN #IMPLIED pattern CDATA #IMPLIED rights NMTOKEN #IMPLIED
    stealth NMTOKEN #IMPLIED value CDATA #IMPLIED>
]>
<!--
  Configure ImageMagick policies.

  Domains include system, delegate, coder, filter, path, or resource.

  Rights include none, read, write, execute and all.  Use | to combine them,
  for example: "read | write" to permit read from, or write to, a path.

  Use a glob expression as a pattern.

  Suppose we do not want users to process MPEG video images:

    <policy domain="delegate" rights="none" pattern="mpeg:decode" />

  Here we do not want users reading images from HTTP:

    <policy domain="coder" rights="none" pattern="HTTP" />

  The /repository file system is restricted to read only.  We use a glob
  expression to match all paths that start with /repository:

    <policy domain="path" rights="read" pattern="/repository/*" />

  Lets prevent users from executing any image filters:

    <policy domain="filter" rights="none" pattern="*" />

  Any large image is cached to disk rather than memory:

    <policy domain="resource" name="area" value="1GP"/>

  Use the default system font unless overwridden by the application:

    <policy domain="system" name="font" value="/usr/share/fonts/favorite.ttf"/>

  Define arguments for the memory, map, area, width, height and disk resources
  with SI prefixes (.e.g 100MB).  In addition, resource policies are maximums
  for each instance of ImageMagick (e.g. policy memory limit 1GB, -limit 2GB
  exceeds policy maximum so memory limit is 1GB).

  Rules are processed in order.  Here we want to restrict ImageMagick to only
  read or write a small subset of proven web-safe image types:

    <policy domain="delegate" rights="none" pattern="*" />
    <policy domain="filter" rights="none" pattern="*" />
    <policy domain="coder" rights="none" pattern="*" />
    <policy domain="coder" rights="read|write" pattern="{GIF,JPEG,PNG,WEBP}" />
-->
<policymap>
  <!-- <policy domain="resource" name="temporary-path" value="/tmp"/> -->
  <policy domain="resource" name="memory" value="256MiB"/>
  <policy domain="resource" name="map" value="512MiB"/>
  <policy domain="resource" name="width" value="16KP"/>
  <policy domain="resource" name="height" value="16KP"/>
  <!-- <policy domain="resource" name="list-length" value="128"/> -->
  <policy domain="resource" name="area" value="128MP"/>
  <policy domain="resource" name="disk" value="1GiB"/>
  <!-- <policy domain="resource" name="file" value="768"/> -->
  <!-- <policy domain="resource" name="thread" value="4"/> -->
  <!-- <policy domain="resource" name="throttle" value="0"/> -->
  <!-- <policy domain="resource" name="time" value="3600"/> -->
  <!-- <policy domain="coder" rights="none" pattern="MVG" /> -->
  <!-- <policy domain="module" rights="none" pattern="{PS,PDF,XPS}" /> -->
  <!-- <policy domain="path" rights="none" pattern="@*" /> -->
  <!-- <policy domain="cache" name="memory-map" value="anonymous"/> -->
  <!-- <policy domain="cache" name="synchronize" value="True"/> -->
  <!-- <policy domain="cache" name="shared-secret" value="passphrase" stealth="true"/> -->
  <!-- <policy domain="system" name="max-memory-request" value="256MiB"/> -->
  <!-- <policy domain="system" name="shred" value="2"/> -->
  <!-- <policy domain="system" name="precision" value="6"/> -->
  <!-- <policy domain="system" name="font" value="/path/to/font.ttf"/> -->
  <!-- <policy domain="system" name="pixel-cache-memory" value="anonymous"/> -->
  <!-- <policy domain="system" name="shred" value="2"/> -->
  <!-- <policy domain="system" name="precision" value="6"/> -->
  <!-- not needed due to the need to use explicitly by mvg: -->
  <!-- <policy domain="delegate" rights="none" pattern="MVG" /> -->
  <!-- use curl -->
  <policy domain="delegate" rights="none" pattern="URL" />
  <policy domain="delegate" rights="none" pattern="HTTPS" />
  <policy domain="delegate" rights="none" pattern="HTTP" />
  <!-- in order to avoid to get image with password text -->
  <policy domain="path" rights="none" pattern="@*"/>
  <!-- disable ghostscript format types -->
  <policy domain="coder" rights="none" pattern="PS" />
  <policy domain="coder" rights="none" pattern="PS2" />
  <policy domain="coder" rights="none" pattern="PS3" />
  <policy domain="coder" rights="none" pattern="EPS" />
  <!-- <policy domain="coder" rights="none" pattern="PDF" /> -->
  <policy domain="coder" rights="none" pattern="XPS" />
</policymap>
```

Add the below as a file policy.xml and add it to the path `/etc/ImageMagick-6/policy.xml.

OddMagnet replied:
Not sure if there's a better way, for now I simply used another `COPY` command in the dockerfile

nuvious commented:
@OddMagnet, if you have time, I left some comments on how to adjust your PR. If you have time to fix them go for it; will give you more credit on the contribution side. If you don't, I can make the changes myself. Always happy to answer any questions.

OddMagnet committed:
Update readme to include docker compose template

OddMagnet commented:
Hey @nuvious, I've updated the package in the Readme and added a template for docker compose (though that'll need to be updated again with the image from docker hub once that's done)

nuvious commented on README.md:
```diff
@@ -107,6 +107,20 @@ EOF
+services:
+  scanpi:
+    container_name: scanpi
+    image: # TODO: Add image here
```
For here, you can use `build: .` to have docker-compose build from the docker file. I will be adding a CICD to publish the container once I have some other issues worked through.

nuvious commented on README.md:
````diff
@@ -107,6 +107,20 @@ EOF
 ```
 
+## Docker Compose Template
+```
````
Add a devices section to the template and fill it with a basic placeholder value. To share a USB device with a docker container it needs to be mapped explicitly.

https://www.homeautomationguy.io/blog/docker-tips/accessing-usb-devices-from-docker-containers

nuvious reviewed (commented):
Great work! If you don't have time to address the comments let me know. If you do address then you'll get more contribution credits and also will help with the learning curve on docker stuff. Happy to answer any questions you have!

OddMagnet committed:
Move installation of dependencies

OddMagnet committed:
Fix policy violation with ImageMagick-6

OddMagnet committed:
Replace 'image:' in compose template with 'build:' for now

OddMagnet committed:
Add device section to compose file with placeholder

OddMagnet commented:
Sorry, I thought you only meant the initial comment, didn't see the others until later 😅 

Pushed the changes for the comments, not sure if another `COPY` command is the best solution for the policy violation problem

fergbrain commented:
Any chance of this merged, or is there more work that needs to be done?

OddMagnet commented:
Gonna close this here if the author is not interested.
